torch/distributed/_composable/contract.py

```diff
@@ -0,0 +1,97 @@
+import torch
+import torch.nn as nn
+
+from collections import OrderedDict
+from typing import Any, Callable, Dict, Optional
+
+
+# use state_slot as key for module.__dict__ to avoid coliding with other
+# properties.
+# TODO: since all composable distributed features can share the same slot.
+class _StateKey:
+
+    # implement operator < to avoid breaking dir()
+    def __lt__(self, other: Any) -> bool:
+        return True if isinstance(other, str) else id(self) < id(other)
+
+
+state_key = _StateKey()
+
+
+def contract(func):
+    def get_all_state(module: nn.Module) -> Dict[Callable, Dict]:
+        d = module.__dict__.setdefault(state_key, {})
+        assert isinstance(d, dict), "Distributed composable API states corrupted"
+        return d
+
+    r"""
+    Decorate a function as a composable distributed API, where the first
+    argument of the function must be an :class:`nn.Module` instance. The
+    decorator verifies that the wrapped function does not modify parameter,
+    buffer, and sub-module fully-qualified names (FQN).
+    """
+
+    def wrapper(module: nn.Module, *args, **kwargs) -> Optional[nn.Module]:
+        # install states specific to the wrapped ``func``
+        all_state: Dict[Callable, dict] = get_all_state(module)
+        assert func not in all_state, (
+            "Each distinct composable distributed API can only be applied to a "
+            f"module once. {func} has already been applied to the following "
+            f"module.\n{module}"
+        )
+        all_state.setdefault(func, {})
+
+        orig_named_params = OrderedDict(module.named_parameters())
+        orig_named_buffers = OrderedDict(module.named_buffers(remove_duplicate=False))
```

Comment:
It seems `module.named_buffers()` has no arg of `remove_duplicate`:
https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.named_buffers